### PR TITLE
Document early TAs

### PR DESCRIPTION
**Pull request by jforissier**
Document early TAs introduced by commit d0c636148b3a ("Add support for
early Trusted Applications") and slightly rework the section about normal
TAs.

Signed-off-by: Jerome Forissier <jerome.forissier@linaro.org>

### CODE REVIEW

File: documentation/optee_design.md [hunk 1]
```diff
@@ -615,53 +614,64 @@ and are built along with and statically built into the OP-TEE core blob.
...
+These pseudo-TAs are used for various purposes such as specific secure services
+or embedded tests services.
 
 Pseudo TAs do not benefit from the GlobalPlatform Core Internal API support
 specified by the GlobalPlatform TEE specs. These APIs are provided to TAs as a
 static library each TA shall link against (the "libutee") and that calls OP-TEE
-core service through system calls. As OP-TEE core does link with the
-libutee, Pseudo TAs can only use the OP-TEE core internal APIs and
-routines.
+core service through system calls. As OP-TEE core does link with the libutee,
```
Comment (jenswi-linaro) [inline documentation/optee_design.md]: doesn't link with

Comment (jforissier) [reply documentation/optee_design.md]: fixed

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: documentation/optee_design.md [hunk 2]
```diff
@@ -615,53 +614,64 @@ and are built along with and statically built into the OP-TEE core blob.
...
+particular application UUID. They run at a lower CPU privilege level
+than OP-TEE core code. In that respect, they are quite similar to regular
+applications running in the Rich Execution Environment (REE), except that
+they execute in Secure World.
 
 Trusted Application benefit from the GlobalPlatform Core Internal API as
 specified by the GlobalPlatform TEE specifications.
 
-Trusted Application consist of a cleartext signed ELF file, named from the UUID
+There are two types of user mode TAs, which differ by the way they are stored.
```
Comment (etienne-lms) [inline documentation/optee_design.md]: proposal:
* There are **~~two~~several** types of user mode TAs, which differ by the way they are stored **~~.~~**
**in the system and get loaded in the execution context created by the OP-TEE core.**

File: documentation/optee_design.md [hunk 3]
```diff
@@ -615,53 +614,64 @@ and are built along with and statically built into the OP-TEE core blob.
...
+
+#### "Normal" or REE FS Trusted Applications
+
+They consist of a cleartext signed ELF file, named from the UUID
```
Comment (etienne-lms) [inline documentation/optee_design.md]: cleartext or cipher as platform can register its own REE FS load hooks. 
So OP-TEE currently proposes cleartext, but can be used to load ciphered signed ELF.

**Comment by etienne-lms** [comment]
merged 16h ago!   i'm really late....

File: documentation/optee_design.md (hunk 2 above)
Comment (jforissier) [reply documentation/optee_design.md]: At the moment there are only two types

File: documentation/optee_design.md (hunk 3 above)
Comment (jforissier) [reply documentation/optee_design.md]: The doc describes what is implemented upstream now, not what can possibly be implemented downstream.
That being said, we could add a paragraph about the TA store.

File: documentation/optee_design.md [hunk 4]
```diff
@@ -615,53 +614,64 @@ and are built along with and statically built into the OP-TEE core blob.
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply documentation/optee_design.md]: I'll take care of this.

**Pull request merged**